coffea/jetmet_tools/CorrectedJetsFactory.py

```diff
@@ -394,63 +404,82 @@ def build(self, jets, lazy_cache):
                     form=scalar_form,
                     cache=lazy_cache,
                 )
-                up[self.name_map["JetMass"]] = awkward.virtual(
+                variation[jetmass] = awkward.virtual(
                     junc_smeared_val,
                     args=(
-                        out[f"jet_energy_uncertainty_{name}"],
-                        0,
-                        out[juncnames["JetMass"]],
+                        unc,
+                        updown,
+                        jetmass_orig,
                     ),
                     length=len(out),
                     form=scalar_form,
                     cache=lazy_cache,
                 )
+                return variation
 
-                down = awkward.flatten(jets)
-                down[self.name_map["JetPt"]] = awkward.virtual(
-                    junc_smeared_val,
-                    args=(
-                        out[f"jet_energy_uncertainty_{name}"],
-                        1,
-                        out[juncnames["JetPt"]],
-                    ),
+            def build_variant(injets, unc, jetpt, jetpt_orig, jetmass, jetmass_orig):
+                up = awkward.virtual(
+                    build_variation,
+                    args=(injets, unc, jetpt, jetpt_orig, jetmass, jetmass_orig, 0),
                     length=len(out),
-                    form=scalar_form,
                     cache=lazy_cache,
                 )
-                down[self.name_map["JetMass"]] = awkward.virtual(
-                    junc_smeared_val,
-                    args=(
-                        out[f"jet_energy_uncertainty_{name}"],
-                        1,
-                        out[juncnames["JetMass"]],
-                    ),
+                down = awkward.virtual(
+                    build_variation,
+                    args=(injets, unc, jetpt, jetpt_orig, jetmass, jetmass_orig, 1),
                     length=len(out),
-                    form=scalar_form,
                     cache=lazy_cache,
                 )
-                out[f"JES_{name}"] = awkward.zip(
+                return awkward.zip(
                     {"up": up, "down": down}, depth_limit=1, with_name="JetSystematic"
                 )
 
+            for name, func in juncs:
+                # out[f"jet_energy_uncertainty_{name}"]
+                out_dict[f"jet_energy_uncertainty_{name}"] = func
+
+                # out[f"JES_{name}"]
+                out_dict[f"JES_{name}"] = awkward.virtual(
+                    build_variant,
+                    args=(
+                        jets,
+                        out_dict[f"jet_energy_uncertainty_{name}"],
+                        self.name_map["JetPt"],
+                        out_dict[juncnames["JetPt"]],
+                        self.name_map["JetMass"],
+                        out_dict[juncnames["JetMass"]],
+                    ),
+                    length=len(out),
+                    cache=lazy_cache,
```

Comment:
Putting a cache here makes me nervous, can you maybe add a check in the tests for proper cleanup if jets are made and then maybe just one thing is accessed?

Reply:
Replaced with non-user cache.

Reply:
I think not specifying cache works too? More safe since `awkward._util` is semi-private